lwt/happy_eyeballs_lwt.ml

```diff
@@ -46,21 +51,30 @@ let try_connect ip port =
 
 let rec act t action =
   let open Lwt.Infix in
+  let ( >>? ) = Lwt_result.bind in
   Log.debug (fun m -> m "[%u] action %a" t.counter
                 Happy_eyeballs.pp_action action);
   begin
     match action with
-    | Happy_eyeballs.Resolve_a host ->
-      begin
-        Dns_client_lwt.getaddrinfo t.dns Dns.Rr_map.A host >|= function
-        | Ok (_, res) -> Ok (Happy_eyeballs.Resolved_a (host, res))
-        | Error `Msg msg -> Ok (Happy_eyeballs.Resolved_a_failed (host, msg))
-      end
-    | Happy_eyeballs.Resolve_aaaa host ->
+    | Happy_eyeballs.Resolve_a host | Happy_eyeballs.Resolve_aaaa host ->
       begin
-        Dns_client_lwt.getaddrinfo t.dns Dns.Rr_map.Aaaa host >|= function
-        | Ok (_, res) -> Ok (Happy_eyeballs.Resolved_aaaa (host, res))
-        | Error `Msg msg -> Ok (Happy_eyeballs.Resolved_aaaa_failed (host, msg))
+        let record = match action with
+          | Happy_eyeballs.Resolve_a _ -> `A
+          | Happy_eyeballs.Resolve_aaaa _ -> `AAAA
+          | _ -> assert false (* never occur! *) in
+        Lwt.return (Domain_name.host host) >>?
+        t.getaddrinfo record >|= fun res -> match res, record with
+        | Ok set, `A ->
+          let fold ip set = match ip with
+            | Ipaddr.V4 ipv4 -> Ipaddr.V4.Set.add ipv4 set
+            | _ -> set in
+          Ok (Happy_eyeballs.Resolved_a (host, Ipaddr.Set.fold fold set Ipaddr.V4.Set.empty))
+        | Ok set, `AAAA ->
+          let fold ip set = match ip with
+            | Ipaddr.V6 ipv6 -> Ipaddr.V6.Set.add ipv6 set
+            | _ -> set in
+          Ok (Happy_eyeballs.Resolved_aaaa (host, Ipaddr.Set.fold fold set Ipaddr.V6.Set.empty))
```

Comment:
so, what is the deal here? we weaken our type of `getaddrinfo`, and now we defer to runtime checking?

I'd propose to either stay on the type level (and have a GADT doing `A of Ipaddr.V4.Set.t | Aaaa of Ipaddr.V6.Set.t`) or just deal with whatever the implementor provided (and use `Ipaddr.Set.union`).

Another option could be to have "getaddrinfo and getaddrinfo6", where the former is `Domain_name.t -> (Ipaddr.V4.Set.t, _) result` - the latter `Domain_name.t -> (Ipaddr.V6.Set.t, _) result` --> no GADT, but at the price of two functions needed to be injected.